sdn: move from EventQueue to DeltaFIFO

dcbw opened the pull request:
The OpenShift controller cache should not be used in new code and is eventually going away.

See https://github.com/openshift/origin/pull/9981#discussion_r71818202

@openshift/networking @danwinship @pravisankar @smarterclayton

danwinship commented on pkg/sdn/plugin/eventqueue.go:
```diff
@@ -0,0 +1,140 @@
+		cache.DeltaCompressorFunc(func(d cache.Deltas) cache.Deltas {
+			return deltaCompressor(d, keyFunc)
+		}),
+		cache.NewStore(keyFunc),
```
This doesn't seem like it should work; the last argument to NewDeltaFIFO() is treated read-only by DeltaFIFO, so by passing an empty Store (which will always remain an empty Store) you are effectively claiming that you never know about any objects, even after being told about them. This should cause you to lose Delete notifications.

danwinship commented on pkg/sdn/plugin/eventqueue.go:
```diff
@@ -0,0 +1,140 @@
+	"k8s.io/kubernetes/pkg/client/cache"
+)
+
+func NewEventQueue(keyFunc cache.KeyFunc) *cache.DeltaFIFO {
```
it seems unlikely that the sdn code is going to be the only part of origin that wants this delta-compressing code...

dcbw replied:
Currently it's just pkg/router/controller/factory/factory.go and that's apparently going to stop using it.  @smarterclayton should I just replace pkg/client/cache/eventqueue.go with this new DeltaFIFO-based event queue, and convert everything over to use it?

smarterclayton replied:
It's only router and SDN, so if you're confident in the correctness of your implementation and we have enough time / effort to properly test it, then I'm ok to go ahead there.

danwinship commented on pkg/sdn/plugin/eventqueue.go:
```diff
@@ -0,0 +1,140 @@
+
+const (
+	// The cache event should result in an add
+	cacheEventEffectAdd cacheEventEffect = "ADD"
```
The naming here is confusing. I think the internal delta list management needs to use different terms than the external event queue (to make it clear that this is about adding an item to the list of notifications, and not necessarily about an item being added to the set the caller is monitoring).

dcbw commented:
@danwinship I fixed for all your comments, now repushed.  I'd like to get this in earlier since it's blocking the CNI plugin stuff, and then perhaps we could make it more generalized and move the router controller factory over to it later?

dcbw commented:
[test]

smarterclayton commented:
Changes look good other than the comment about replacing the router code as well.

dcbw commented:
> Changes look good other than the comment about replacing the router code as well.

@smarterclayton thanks; is it your intention to move the router code away from NewEventQueue, or just keep it private to the router code?

pravisankar commented:
LGTM

smarterclayton commented:
Goa is to replace router with an informer + indexer and use the events
to trigger recalculation.  However, needs to be worked through in more
detail.

dcbw commented:
@openshift/networking @danwinship @pravisankar back up for review; but we may want to *not* merge this until the freeze is done.

dcbw commented:
Flake:

```
++ Building go targets for linux/amd64: cmd/openshift cmd/oc

# github.com/openshift/origin/vendor/google.golang.org/grpc/transport

vendor/google.golang.org/grpc/transport/http_util.go:271: internal compiler error: schedule does not include all values
```

dcbw commented:
[test]

dcbw commented:
flake is https://github.com/openshift/origin/issues/10445

dcbw commented:
re-[test]

dcbw commented:
re-[test] flake #10445

dcbw commented:
re-[test] now that travis seems happier

dcbw commented:
re-[test] flake https://github.com/openshift/origin/issues/9938

dcbw commented:
Updated for a bug; Deleted events wouldn't be propagated when the queue was empty for an object, due to how DeltaFIFO tracks things internally when knownObjects is nil.  Thus we must implement our own store in EventQueue to get reliable deletion notification.

dcbw commented:
jenkins failure is flake https://github.com/openshift/origin/issues/10080

dcbw commented:
[test]

danwinship commented:
[merge]
